File: integrations/tests/tests/unit/test_dataset_name_conformance.py
"""Conformance: the hook plugins' dataset-name sanitizer matches the shared spec.

``integrations/conformance/dataset_name_cases.json`` is the single source of truth
for the rule every integration applies to a dataset name: cognee rejects names
containing a space or a dot (``check_dataset_name``, run on every write), so
those — and only those — are rewritten. Each integration checks its own
sanitizer against the one table.

Beyond the table: a configured ``COGNEE_PLUGIN_DATASET`` is sanitized when the
config loads, an explicit switch to an invalid name is refused rather than
rewritten, and code-graph dataset names never carry a dot.
"""

from __future__ import annotations

import json
from pathlib import Path

import pytest
from utils.suites import ANTIGRAVITY, CLAUDE, CODEX

_PLUGINS = [CLAUDE, CODEX, ANTIGRAVITY]


def _load_cases() -> list[dict]:
    integrations = next(p for p in Path(__file__).resolve().parents if p.name == "integrations")
    cases = integrations / "conformance" / "dataset_name_cases.json"
    return json.loads(cases.read_text(encoding="utf-8"))


@pytest.mark.parametrize("plugin_suite", _PLUGINS, ids=lambda s: s.name)
def test_sanitizer_matches_shared_table(plugin_suite, isolated_modules):
    config = isolated_modules(plugin_suite, "config")

    mismatches = []
    for case in _load_cases():
        result = config.sanitize_dataset_name(case["input"], case["fallback"])
        if result != case["expected"]:
            mismatches.append(f"{case['input']!r} -> {result!r}, expected {case['expected']!r}")

    assert not mismatches, "dataset-name sanitization drift:\n" + "\n".join(mismatches)


@pytest.mark.parametrize("plugin_suite", _PLUGINS, ids=lambda s: s.name)
def test_configured_dataset_is_sanitized_and_logged(plugin_suite, isolated_modules, monkeypatch):
    config = isolated_modules(plugin_suite, "config")
    events: list[tuple[str, dict]] = []
    monkeypatch.setattr(
        config, "_config_log", lambda event, detail=None: events.append((event, detail))
    )

    monkeypatch.setenv("COGNEE_PLUGIN_DATASET", "my project.v2")
    assert config.load_config()["dataset"] == "my_project_v2"
    assert events == [("dataset_name_sanitized", {"from": "my project.v2", "to": "my_project_v2"})]

    events.clear()
    monkeypatch.setenv("COGNEE_PLUGIN_DATASET", "Foo+Bar")
    assert config.load_config()["dataset"] == "Foo+Bar"
    assert events == []


@pytest.mark.parametrize("plugin_suite", _PLUGINS, ids=lambda s: s.name)
def test_switch_refuses_an_invalid_name_with_a_suggestion(plugin_suite, hook_module):
    switch = hook_module(plugin_suite, "switch-dataset.py")
    with pytest.raises(
        switch.SwitchError, match=r"cognee rejects spaces and dots \(try 'my_data'\)"
    ):
        switch._switch("host-1", {}, "my.data", force=False)
    with pytest.raises(switch.SwitchError, match="cognee rejects spaces and dots$"):
        switch._switch("host-1", {}, "...", force=False)


@pytest.mark.parametrize("plugin_suite", _PLUGINS, ids=lambda s: s.name)
def test_code_graph_dataset_name_has_no_dot(plugin_suite, isolated_modules, tmp_path):
    code_graph = isolated_modules(plugin_suite, "_code_graph")
    repo = tmp_path / "Foo.JS"
    repo.mkdir()
    name = code_graph.default_code_dataset(str(repo))
    assert name.startswith("codebase-foo-js-")
    assert "." not in name

File: integrations/openclaw/__tests__/unit/test_dataset_name_conformance.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";

import { defaultCodeDataset } from "../../src/code-graph";
import { resolveConfig } from "../../src/config";
import { datasetNameForScope, sanitizeDatasetName } from "../../src/scope";

type SanitizeCase = { input: string; fallback: string; expected: string; note: string };

// The shared case table lives at integrations/conformance/dataset_name_cases.json.
// The claude-code, codex, antigravity and hermes-agent tests read the same file,
// so any implementation that drifts from the shared rule fails its test.
const casesPath = resolve(__dirname, "..", "..", "..", "conformance", "dataset_name_cases.json");
const cases: SanitizeCase[] = JSON.parse(readFileSync(casesPath, "utf-8"));

describe("dataset-name sanitization conformance", () => {
  it.each(cases)("$note", ({ input, fallback, expected }) => {
    expect(sanitizeDatasetName(input, fallback)).toBe(expected);
  });
});

describe("dataset names reach cognee sanitized", () => {
  const saved = process.env.COGNEE_PLUGIN_DATASET;
  afterEach(() => {
    if (saved === undefined) delete process.env.COGNEE_PLUGIN_DATASET;
    else process.env.COGNEE_PLUGIN_DATASET = saved;
  });

  it("sanitizes the configured dataset name", () => {
    process.env.COGNEE_PLUGIN_DATASET = "my project.v2";
    expect(resolveConfig({}).datasetName).toBe("my_project_v2");
  });

  it("leaves a name the server accepts unchanged", () => {
    process.env.COGNEE_PLUGIN_DATASET = "Foo+Bar";
    expect(resolveConfig({}).datasetName).toBe("Foo+Bar");
  });

  it("sanitizes names derived from user and agent ids", () => {
    delete process.env.COGNEE_PLUGIN_DATASET;
    const cfg = resolveConfig({ datasetName: "team", userId: "a.b@x.com" });
    expect(datasetNameForScope("user", cfg)).toBe("team-user-a_b@x_com");
  });

  it("gives code-graph datasets a dot-free name", () => {
    const name = defaultCodeDataset("/work/Foo.JS");
    expect(name.startsWith("codebase-foo-js-")).toBe(true);
    expect(name).not.toContain(".");
  });
});
